lib/yarp.rb

```diff
@@ -165,8 +167,64 @@ def visit_child_nodes(node)
   # This lexes with the Ripper lex. It drops any space events but otherwise
   # returns the same tokens.
   def self.lex_ripper(source)
-    Ripper.lex(source).each_with_object([]) do |token, tokens|
-      tokens << token unless token[1] == :on_sp
+    Ripper.lex(source).each_with_object([]) do |token_arr, tokens|
+      token = LexedToken.new(*token_arr)
+      tokens << token unless token.event == :on_sp
+    end
+  end
+
+  class LexedToken < SimpleDelegator
+    attr_reader :location, :event, :value
+    attr_accessor :state
+
+    def initialize(location, event, value, state)
+      @location = location
+      @event = event
+      @value = value
+      @state = state
+    end
+
+    def [](index)
+      inspect[index]
+    end
+
+    def ==(other)
+      other in LexedToken[
+        location: ^(location),
+        event: ^(event),
+        value: ^(value),
+        state: ^(state)
+      ] or other in LexedToken[
+        location: ^(location),
+        event: :on_comment,
+        value: ^(value)
+      ]
+    end
+
+    def is_tilde_heredoc?
+      value[2] == "~"
+    end
+
+    def inspect
+      [@location, @event, @value, @state]
+    end
+
+    def deconstruct_keys(keys)
+      { location: location, event: event, value: value, state: state }
+    end
+
+    def pretty_print(q)
```

Comment:
I'm new to `pretty_print` so copied other methods, let me know if this should be done differently though!